### PR TITLE
VE-141 - Re-implement NumberField

### PR DESCRIPTION
**Pull request by lukemojo**
Contributes [TD-141](https://sce.myjetbrains.com/youtrack/issue/VE-141/Re-implement-string-conversion-to-number-in-NumberField)

## Changes

Reinstates all the previous functionality of NumberField along with the string to number conversion. Now returns a value rather than an event. This is a breaking change requiring updates in the VIRTO and CRA apps.

There is a sneaky one liner in here fixing the colour of the ConfirmDialog text in dark mode.

A brief description of what this pull request solves / introduces.

- This effectively reverts the change to using a standard MUI Input to performing number conversion
- The output has changed because this is now a TS component so we no longer try and hijack the event 
- These changes prevent the need for converting to a number in applications (but will still require string coercion in Zod session schemas)
- Reinstating the old functionality should allow instrument designer updates with minimum updates
- Fixes dark mode confirmation text

## Testing notes

- NumberField stories should now show a value onChange rather than the event
- Text should be white in dark mode ConfirmDialog

## Author checklist

Before I request a review:

- [x] I have reviewed my own code-diff.
- [x] I have tested a pre-release
- [x] I have assigned the PR to myself or an appropriate delegate.
- [x] I have added the relevant labels to the PR.
- [x] I have included appropriate tests.
- [x] I have checked that the Lint and Test workflows pass on Github.
- ~~[ ] I have populated the deploy-preview with relevant test data.~~

### CODE REVIEW

**Comment by mattcorner** [comment]
Shouldn't RUI PRs have a target release? This is currently destined for main.

**Review by mattcorner — CHANGES_REQUESTED** [review]
Discussed to undo the breaking change to the onChange interface and correctly type it. Applications can use `NumberFieldProps["onChange"]` to get the correct types for the callback method.

We should be much more considerate about making breaking changes in components like this that are widely used not just in one or two VIRTO apps, but across the whole suite as well as in other tools.

File: src/NumberField/NumberField.test.tsx [hunk 1]
```diff
@@ -14,13 +14,13 @@ const NumberFieldWithState = ({
   valueIn = 100,
   ...rest
 }: {
-  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
-  valueIn?: number;
+  onChange?: (value: number) => void;
```
Comment (mattcorner) [inline src/NumberField/NumberField.test.tsx:17]: Is this correct? onChange takes an event. Why do we even need to override the type for this?

File: src/NumberField/NumberField.tsx [hunk 2]
```diff
@@ -52,6 +119,58 @@ export default function NumberField(props: NumberFieldProps) {
...
       type="number"
+      value={currentValue}
     />
   );
 }
+
+/**
+ * Converts a string to a number.
+ * @param {*} value - The value to convert.
+ * @returns {number} - The converted value.
```
Comment (mattcorner) [inline src/NumberField/NumberField.tsx:130]: Don't add types to jsdoc
```suggestion
 * @param value - The value to convert.
 * @returns - The converted value (either a number of null)
```

Comment (mattcorner) [reply src/NumberField/NumberField.tsx:130]: Same point applies to all of these utility methods.

File: src/NumberField/NumberField.stories.tsx [hunk 3]
```diff
@@ -18,9 +18,7 @@ const Template: StoryFn<NumberFieldProps> = args => {
     setValue(args.value);
   }, [args.value]);
 
-  const onChange = (
-    event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
-  ) => {
+  const onChange = (event: NumberFieldChangeEvent) => {
```
[the remark this answers is not in the document]
Comment (mattcorner) [reply src/NumberField/NumberField.stories.tsx:21]: As a result I'd prefer not to export NumberFieldChangeEvent from RUI. This is consistent with MUI.

File: src/NumberField/NumberField.stories.tsx [hunk 4]
```diff
@@ -157,8 +164,31 @@ export const InitialError: Story = {
     helperText: "Oh no you've made a mistake!",
     label: "Enter a Number",
     margin: "normal",
+    max: 1,
```
Comment (mattcorner) [inline src/NumberField/NumberField.stories.tsx:167]: You've set the initial value on this story to something that shows an error message other than the one passed in. Suggest that you set the initial value to something valid / don't include a min/,ax for this story.

File: src/NumberField/NumberField.types.ts [hunk 5]
```diff
@@ -1,11 +1,36 @@
...
+  | "minRows"
+  | "multiline"
+  | "onChange"
+  | "rows"
+  | "select"
+  | "sx"
+  | "type"
+> & {
   endAdornment?: string;
+  min?: number;
```
Comment (mattcorner) [inline src/NumberField/NumberField.types.ts:28]: Missing jsdoc for these props

File: src/NumberField/NumberField.stories.tsx
Comment (mattcorner) [inline src/NumberField/NumberField.stories.tsx:1]: Only some of the stories here are showing the onBlur action. I thought this should come for free so a bit confused why we handle it manually in some and not for others.

**Review by mattcorner — CHANGES_REQUESTED** [review]

**Comment by dmbarry86** [comment]
> Shouldn't RUI PRs have a target release? This is currently destined for main.

@lukemojo we've had to do v8.10.0 release whilsy you were not working today so you will need to target another release number now for this PR. Still it is targetted to main also. I'll leave you to create the next release branch.

**Review by mattcorner — CHANGES_REQUESTED** [review]
Is there a corresponding VIRTO branch where we have proven that the pre-release for this fixes [TD-3085](https://sce.myjetbrains.com/youtrack/issue/TD-3085/Error-publishing-a-sensor) and also integrates into the wizards e.g. SCENE?

File: src/NumberField/NumberField.test.tsx (hunk 1 above)
Comment (mattcorner) [reply src/NumberField/NumberField.test.tsx]: Hasn't been addressed.

**Comment by lukemojo** [comment]
> Is there a corresponding VIRTO branch where we have proven that the pre-release for this fixes [TD-3085](https://sce.myjetbrains.com/youtrack/issue/TD-3085/Error-publishing-a-sensor) and also integrates into the wizards e.g. SCENE?

There is a PR for this in VIRTO that I haven't yet linked as it's on the pre-release. 
https://github.com/IPG-Automotive-UK/virto/pull/1094

If we're happy with how the deploy-preview is functioning then I will release this and update in the VIRTO PR

**Review by mattcorner — APPROVED** [review]
LGTM. Tested in https://github.com/IPG-Automotive-UK/virto/pull/1094 and works well. Go ahead and release and integrate into that PR.

**Pull request merged**